python/lsst/source/injection/inject_engine.py

```diff
@@ -444,8 +453,38 @@ def inject_galsim_objects_into_exposure(
     bbox = exposure.getBBox()
     full_bounds = galsim.BoundsI(bbox.minX, bbox.maxX, bbox.minY, bbox.maxY)
     galsim_image = galsim.Image(exposure.image.array, bounds=full_bounds)
+    galsim_variance = galsim.Image(exposure.variance.array, bounds=full_bounds)
     pixel_scale = wcs.getPixelScale(bbox.getCenter()).asArcseconds()
 
+    variance_scale = galsim_variance.copy()
+
+    # Figure out if this exposure has well-defined amplifier regions with gains
+    try:
+        amplifiers = exposure.getDetector().getAmplifiers()
+        amp_bboxes = [amplifier.getBBox() for amplifier in amplifiers]
+        gains = [amplifier.getGain() for amplifier in amplifiers]
+    except AttributeError:
+        gains = None
+
+    if gains is None:
+        # Gain info has not been stored in the exposure, so estimate it with a
+        # linear fit of image flux vs. variance in each pixel.
+        image_arr = exposure.image.array
+        var_arr = exposure.variance.array
+        mask_arr = exposure.mask.array
+        good = np.isfinite(image_arr) & np.isfinite(var_arr)
+        bad_mask_names = ["BAD", "CR", "CROSSTALK", "INTRP", "NO_DATA", "SAT", "SUSPECT", "UNMASKEDNAN"]
+        maskPlaneDict = exposure.mask.getMaskPlaneDict()
+        for mask_name in bad_mask_names:
+            if mask_name in maskPlaneDict:
+                good &= (mask_arr & (1 << maskPlaneDict[mask_name])) == 0
+        fit = np.polyfit(image_arr[good], var_arr[good], deg=1)
```

Comment:
Here, we're fitting the entire exposure to get an estimate of gain. However, below (if `gains`), we're constructing `variance_scale` per amp. Is there a good reason not to perform the image <-> variance fit per amp as well? Or is a small numbers statistics problem that you're worried about?

Reply:
The discussion below clarifies this, but for the sake of completeness I'll say here that the intention is only to fit the image flux vs. variance if it's a coadd (which doesn't have distinct amplifier regions).

Reply:
Thank you. New logic to make clear if we're dealing with a `calexp` or not is also helpful.